### PR TITLE
Make TypeBuilder.CreateType return non nullable

### LINKED ISSUE #68840: Under what conditions can `TypeBuilder.CreateType()` return null
**Issue description**
Under what conditions can `TypeBuilder.CreateType()` return null?

**Target framework**
- [x] .NET Core 5, 6
- [ ] .NET Framework
- [ ] .NET Standard

**Details**
The `TypeBuilder.CreateType()` method is defined [as nullable](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.typebuilder.createtype#system-reflection-emit-typebuilder-createtype):
````cs
public Type? CreateType();
````

Under what conditions can it return null? The docs do not say.

Please document this method signature.

### PR DESCRIPTION
**Pull request by JosieBigler**
There is no situation when TypeBuilder.CreateType will return null when
being called by user-code.

Fixes #68840

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=72180) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=72180) before we can accept your contribution.<br/><br/>:x: JosieBigler [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=72180)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=72180) it.</sub>

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs [hunk 1]
```diff
@@ -694,6 +694,17 @@ private void SetGenParamCustomAttributeNoLock(CustAttr ca)
             m_ca ??= new List<TypeBuilder.CustAttr>();
             m_ca.Add(ca);
         }
+
+        internal void CreateGlobalModuleType()
```
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:698]: Who's using this?

Comment (JosieBigler) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:698]: I implemented the changes outlined by @Jehoel in the [issue](https://github.com/dotnet/runtime/issues/68840).  

> > @lonix1 Under what conditions can `TypeBuilder.CreateType()` return null?
> 
> **TL;DR: There is no situation** when `TypeBuilder.CreateType()` will return `null` when `CreateType()` is being called by user-code. You can safely add a null-forgiving `!` to a `CreateType()` call-site.
> 
> The current source is at
> 
> https://github.com/dotnet/runtime/blob/1a296c06fff8750b4658f6f7d901347e006744d0/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs#L1853
> 
>     * A quick browse of the current source  shows that `TypeBuilder.CreateType()` only returns `null` when the `TypeBuilder` is the builder for an assembly's hidden `<Module>` class.
>       
>       * The `<Module>` class is a special type that allows assemblies/modules to have eager-initialized globals.
>       * The C# language itself doesn't use `<Module>`, though other languages do (e.g. C++/CLI).
>       * Since C# 9.0, [you _can_ now opt-in to adding `<Module>` members](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/module-initializers) via the `[ModuleInitializer]` attribute.
>       * Of course, if you're using `Reflection.Emit` then you will have always been able to use `ModuleBuilder` going back to .NET 1.x in 2001.
> 
>     * `<Module>`-building `TypeBuilder` instances cannot be directly instantiated by user-code as it requires a specific `internal TypeBuilder()` constructor invocation - instead this ctor is called only when you use the `ModuleBuilder` type.
> 
>     * The only situation when `TypeBuilder.CreateType()` method _does_ indeed return `null` is inside `void ModuleBuilder.CreateGlobalFunctions()`, which basically ensures that a `class <Module>` _will_ be defined in the generated dynamic-assembly.
> 
> 
> So based on that, I think the design of `TypeBuilder`'s API should be changed in a minimally-breaking way:
> 
>     * Change the current `public Type? TypeBuilder.CreateType()` method into `private Type? TypeBuilder.CreateTypeImpl()`.
> 
>     * Add back `public Type TypeBuilder.CreateType()` (without the nullable return type) which just calls `CreateTypeImpl()` directly, maybe adding a debug assertion that `CreateTypeImpl` _did not_ return `null`?).
> 
>     * Add a new method to `TypeBuilder`: `internal void CreateGlobalModuleType()` which first checks that `this.m_isHiddenGlobalType == true` (and if not throws, I suppose) and then calls `CreateTypeImpl()` with a debug assertion that it _did_ return `null`.
> 
> 
> Alternatively, refactor the `CreateType()` method more thoroughly?

Comment (stephentoub) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:698]: > I implemented the changes outlined by @Jehoel in the https://github.com/dotnet/runtime/issues/68840.

But presumably the intention is that there's some existing caller of CreateType that would need to be changed to use CreateGlobalModuleType?

Comment (buyaa-n) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:698]: Yes, but not only that, the `CreateGlobalModuleType()` method unconditionally throws so the call never gonna succeed.

Comment (JosieBigler) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:698]: Ah, I've fixed the `throw` in `CreateGlobalModuleType()` and replaced the method call for `_globalTypeBuilder`.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs [hunk 2]
```diff
@@ -694,6 +694,17 @@ private void SetGenParamCustomAttributeNoLock(CustAttr ca)
...
+        {
+            if (!m_isHiddenGlobalType)
+            {
+                throw new NotSupportedException(SR.NotSupported_DynamicModule);
+            }
+
+            Debug.Assert(CreateTypeImpl() == null);
```
Comment (buyaa-n) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:705]: Now this gonna run only for Debug build, I believe its needed for any build.

Anyway, I don't think asserting and checking `m_isHiddenGlobalType` is necessary, so lets not add this extra method, and call `CreateTypeImpl()` directly instead  https://github.com/dotnet/runtime/blob/a712828583b54a283d0e5343457d4b16d2234b31/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.cs#L855

Also the ref assembly need to be updated (remove `?`), plus add a unit test if there is no one exist that covering this scenario

Comment (JosieBigler) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs]: I have made your requested changes. I didn't see an obvious way to test the change to `_globalTypeBuilder.CreateTypeImpl();` in `ModuleBuilder.cs`.

Comment (buyaa-n) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs]: Right, `_globalTypeBuilder.CreateTypeImpl()` is not accessible directly, might add `Assert.Null(member.DeclaringType);`  into tests for global members like https://github.com/dotnet/runtime/blob/main/src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderCreateGlobalFunctions.cs or https://github.com/dotnet/runtime/blob/main/src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderDefineInitializedData.cs, but it is not necessary

File: src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderCreateType.cs [hunk 3]
```diff
@@ -28,6 +28,7 @@ public void CreateType(TypeAttributes attributes)
 
             TypeInfo typeInfo = type.CreateTypeInfo();
             Assert.Equal(typeInfo, createdType.GetTypeInfo());
+            Assert.NotNull(createdType);
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderCreateType.cs:31]: This better to be added before using the `createdType`, just after creating it

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs [hunk 4]
```diff
@@ -694,6 +694,7 @@ private void SetGenParamCustomAttributeNoLock(CustAttr ca)
             m_ca ??= new List<TypeBuilder.CustAttr>();
             m_ca.Add(ca);
         }
+
```
Comment (buyaa-n) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:697]: NIT: avoid unneeded changes
```suggestion
```

File: src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderCreateType.cs (hunk 3 above)
Comment (JosieBigler) [reply src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderCreateType.cs]: Moved the Assertion to just after `createdType` creation.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs (hunk 4 above)
Comment (JosieBigler) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs]: removed extraneous new line.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs (hunk 2 above)
Comment (JosieBigler) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs]: Added Assert.Null to both suite of unit tests after each call of `module.CreateGlobalFunctions();`

**Comment by stephentoub** [comment]
Should EnumBuilder.CreateType be annotated differently as well?

What about CreateTypeInfo?

Can we remove the `!` from https://github.com/dotnet/runtime/blob/0c32267664e1b9913f2329827d998646fecafd91/src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs#L312?

**Comment by JosieBigler** [comment]
@stephentoub  I've removed the `!` from `DispatchProxyGenerator.cs`. 

It looks like we can change `CreateTypeInfo` .  There are a few more null-forgiving operators associated with `CreateTypeInfo` mostly in the `System.Private.Xml` Do you want me to remove the `!` from there as well? Or Should these changes stay confined to `System.Reflection`?

File: src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderCreateGlobalFunctions.cs [hunk 5]
```diff
@@ -17,6 +17,7 @@ public void CreateGlobalFunctions_SingleGlobalMethod()
             ilGenerator.Emit(OpCodes.Ret);
 
             module.CreateGlobalFunctions();
+            Assert.Null(method.DeclaringType);
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderCreateGlobalFunctions.cs:20]: The asserts added to this file fails on mono, https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72180-merge-437d30d8adc24f8fb1/System.Reflection.Emit.Tests/1/console.5d48a2e1.log?helixlogtype=result

Mono implementation looks quite different, seems its better revert these asserts, the same happening 
for Asserts in `ModuleBuilderDefineInitializedData.cs`

Comment (jkotas) [reply src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderCreateGlobalFunctions.cs:20]: Looks like Mono bugs. An issue should be opened on this and the asserts can be kept under `if (!PlatformDetection.IsMonoRuntime) // [ActiveIssue("https://github.com/dotnet/runtime/issues/xxxxxx")]`

**Comment by jkotas** [comment]
> There are a few more null-forgiving operators associated with CreateTypeInfo mostly in the System.Private.Xml Do you want me to remove the ! from there as well?

Yes, you can remove the no-unnecessary `!` from everywhere. Also, you should be able to delete the unnecessary `AsType` calls: replace `CreateTypeInfo()!.AsType()` with just `CreateType()`.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/EnumBuilder.cs [hunk 6]
```diff
@@ -46,7 +46,7 @@ public FieldBuilder DefineLiteral(string literalName, object? literalValue)
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/EnumBuilder.cs:46]: Also change `public TypeInfo? CreateTypeInfo()` to `public TypeInfo CreateTypeInfo()`

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs [hunk 7]
```diff
@@ -1686,7 +1686,15 @@ private EventBuilder DefineEventNoLock(string name, EventAttributes attributes,
         }
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:1686]: Also change `public TypeInfo? CreateTypeInfo()` to 
```
        [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
        public TypeInfo CreateTypeInfo()
        {
            Type? type = CreateTypeImpl();
            Debug.Assert(type != null);
            return type;
        }
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/EnumBuilder.cs (hunk 6 above)
Comment (JosieBigler) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/EnumBuilder.cs:46]: This has been updated.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs (hunk 7 above)
Comment (JosieBigler) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:1694]: I updated this to mimic what was done for `CreateType()`, which amounts to a public API that returns a non-null `TypeInfo`.

File: src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderCreateGlobalFunctions.cs (hunk 5 above)
Comment (JosieBigler) [reply src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderCreateGlobalFunctions.cs:20]: Issue opened, and Failing tests on Mono annotated. https://github.com/dotnet/runtime/issues/72858

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs [hunk 8]
```diff
@@ -1677,7 +1677,15 @@ private EventBuilder DefineEventNoLock(string name, EventAttributes attributes,
...
-        public TypeInfo? CreateTypeInfo()
+        public TypeInfo CreateTypeInfo()
+        {
+            TypeInfo? typeInfo = CreateTypeInfoImpl();
+            Debug.Assert(typeInfo != null);
+            return typeInfo;
+        }
+
+        [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
+        internal TypeInfo? CreateTypeInfoImpl()
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:1688]: CreateTypeInfoImpl is identical to CreateTypeImpl. Could you please delete CreateTypeImpl and make CreateTypeInfo to call CreateTypeImpl instead?

**Comment by jkotas** [comment]
Could you please merge from main to resolve conflict in TypeBuilderAddInterfaceImplementation.cs?

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/IlGen/XmlILModule.cs [hunk 9]
```diff
@@ -239,7 +239,7 @@ public void BakeMethods()
 
             if (!_useLRE)
             {
-                typBaked = _typeBldr!.CreateTypeInfo()!.AsType();
+                typBaked = _typeBldr.CreateType();
```
Comment (jkotas) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/IlGen/XmlILModule.cs:242]:
```suggestion
                typBaked = _typeBldr!.CreateType();
```
I think this needs to stay.

Comment (JosieBigler) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/IlGen/XmlILModule.cs]: Added back the ! to `_typeBldr`.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs (hunk 8 above)
Comment (JosieBigler) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs]: Changed to only have 1 internal method with a signature of `internal TypeInfo? CreateTypeInfoImpl()`.

File: src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderCreateGlobalFunctions.cs (hunk 5 above)
Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderCreateGlobalFunctions.cs:21]: @JosieBigler we want the tests keep running on mono except the new Assert, please move only the `Assert.Null(method.DeclaringType);` part into `if (!PlatformDetection.IsMonoRuntime)` conditional:
```suggestion
        if (!PlatformDetection.IsMonoRuntime) // [ActiveIssue("https://github.com/dotnet/runtime/issues/xxxxxx")]
        {
            Assert.Null(method.DeclaringType);
        }
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs [hunk 10]
```diff
@@ -1677,16 +1677,23 @@ private EventBuilder DefineEventNoLock(string name, EventAttributes attributes,
...
+        public Type CreateType()
+        {
+            Type? type = CreateTypeInfoImpl();
+            Debug.Assert(type != null);
+            return type;
         }
 
         [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
-        public Type? CreateType()
+        internal TypeInfo? CreateTypeInfoImpl()
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:1696]: The naming along the callchain looks a bit odd: `CreateType` -> `CreateTypeInfoImpl` -> `CreateTypeNoLock`

It may be nice to rename `TypeInfo? CreateTypeInfoImpl` to `RuntimeType? CreateRuntime`, and `TypeInfo? CreateTypeNoLock` to `RuntimeType? CreateRuntimeTypeNoLock`

Comment (JosieBigler) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:1696]: When I change from `TypeInfo?` to `RunetimeType?` The `this` at https://github.com/dotnet/runtime/blob/f4cb6ac85acbbbbcdc48893fea16d89fe898e669/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs#L1758  

I get a compile time error: 

`CS00029: Cannot implicitly convert type 'System.Reflection.Emit.TypeBuilder' to 'System.Runtime.Type'`

Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:1696]: Ok, I have not realized this detail. Let's keep it as is then.

**Review by jkotas — APPROVED** [review]
LGMT. Thank you!

**Comment by lonix1** [comment]
@JosieBigler You never thought it would get this complicated when you volunteered... :smile: Seriously though, well done and thanks!
